File: apps/android/app/src/main/java/fr/juliette/thecode/vault/Transfer.java
package fr.juliette.thecode.vault;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.json.JSONException;

import java.io.ByteArrayOutputStream;
import java.nio.charset.StandardCharsets;
import java.security.GeneralSecurityException;
import java.security.SecureRandom;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.zip.Deflater;
import java.util.zip.Inflater;
import java.util.zip.DataFormatException;

import javax.crypto.Cipher;
import javax.crypto.SecretKey;
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.PBEKeySpec;
import javax.crypto.spec.SecretKeySpec;

/**
 * Chiffrement du carnet avant qu'il ne quitte l'appareil.
 *
 * Le serveur ne reçoit que des blocs opaques : il ne peut lire ni les sites, ni
 * les identifiants.
 *
 * Le sel diffère de celui des mots de passe et de celui de l'empreinte : une
 * même valeur dérivée ne doit jamais servir à deux usages, sinon une faiblesse
 * sur l'un exposerait l'autre.
 *
 * Spécification : shared/spec/vault-transfer.md
 */
public final class Transfer {

    public static final String PREFIX = "TC1";

    private static final byte[] SALT = "thecode-transfer/v1".getBytes(StandardCharsets.UTF_8);
    private static final int ITERATIONS = 600_000;
    private static final int KEY_BITS = 256;
    private static final int NONCE_BYTES = 12;
    private static final int TAG_BITS = 128;

    private static final SecureRandom RANDOM = new SecureRandom();

    private Transfer() {}

    /** Dérive la clef de transfert depuis la clef maîtresse. */
    public static SecretKey deriveKey(@NonNull String masterKey) throws GeneralSecurityException {
        PBEKeySpec spec = new PBEKeySpec(masterKey.toCharArray(), SALT, ITERATIONS, KEY_BITS);
        byte[] raw = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
                .generateSecret(spec).getEncoded();
        return new SecretKeySpec(raw, "AES");
    }

    /** Un bloc chiffré : nonce et données, tous deux opaques pour le serveur. */
    public static final class Sealed {
        public final byte[] nonce;
        public final byte[] blob;

        Sealed(byte[] nonce, byte[] blob) {
            this.nonce = nonce;
            this.blob = blob;
        }
    }

    public static Sealed seal(@NonNull SecretKey key, @NonNull String plain)
            throws GeneralSecurityException {
        // Un nonce jamais réutilisé avec la même clef : le contraire casse
        // AES-GCM.
        byte[] nonce = new byte[NONCE_BYTES];
        RANDOM.nextBytes(nonce);

        Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
        cipher.init(Cipher.ENCRYPT_MODE, key, new GCMParameterSpec(TAG_BITS, nonce));
        return new Sealed(nonce, cipher.doFinal(plain.getBytes(StandardCharsets.UTF_8)));
    }

    /**
     * Déchiffre en octets.
     *
     * Le contenu d'un transfert est compressé, donc binaire : passer par une
     * String UTF-8 corromprait les octets qui ne forment pas du texte valide.
     */
    public static byte[] openBytes(@NonNull SecretKey key, @NonNull byte[] nonce,
                                   @NonNull byte[] blob) throws GeneralSecurityException {
        Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
        cipher.init(Cipher.DECRYPT_MODE, key, new GCMParameterSpec(TAG_BITS, nonce));
        return cipher.doFinal(blob);
    }

    /** Déchiffre du texte. Pour du binaire, voir {@link #openBytes}. */
    public static String open(@NonNull SecretKey key, @NonNull byte[] nonce, @NonNull byte[] blob)
            throws GeneralSecurityException {
        return new String(openBytes(key, nonce, blob), StandardCharsets.UTF_8);
    }

    // --------------------------------------- decoupage en plusieurs QR

    /**
     * Un QR plafonne a ~2,9 Ko. On garde de la marge pour l'en-tete du
     * fragment, qui s'ajoute a chaque morceau.
     */
    public static final int FRAGMENT_LIMIT = 2600;

    /**
     * Decoupe un payload en fragments affichables l'un apres l'autre.
     *
     * Un seul fragment quand le payload tient : inutile d'imposer un
     * assemblage pour un carnet ordinaire.
     */
    public static List<String> fragments(@NonNull String payload) {
        List<String> out = new ArrayList<>();
        if (payload.length() <= FRAGMENT_LIMIT) {
            out.add(payload);
            return out;
        }

        // Le prefixe « TC1. » est porte une fois par le reassemblage, pas par
        // chaque fragment.
        String body = payload.substring(PREFIX.length() + 1);
        int total = (body.length() + FRAGMENT_LIMIT - 1) / FRAGMENT_LIMIT;
        for (int i = 0; i < total; i++) {
            int from = i * FRAGMENT_LIMIT;
            int to = Math.min(from + FRAGMENT_LIMIT, body.length());
            out.add("TC1m." + i + "." + total + "." + body.substring(from, to));
        }
        return out;
    }

    /**
     * Assemble les fragments lus. Rend null tant qu'il en manque.
     *
     * L'ordre n'a pas d'importance et un fragment lu deux fois est ignore :
     * les codes defilent en boucle, on ne maitrise pas ce qui est vu quand.
     */
    @Nullable
    public static String assemble(@NonNull Map<Integer, String> fragments, int total) {
        if (total <= 0 || fragments.size() != total) return null;

        StringBuilder body = new StringBuilder();
        for (int i = 0; i < total; i++) {
            String part = fragments.get(i);
            if (part == null) return null;
            body.append(part);
        }
        return PREFIX + "." + body;
    }

    // ------------------------------------------------- carnet entier

    /** Payload illisible : version inconnue, format casse, ou mauvaise clef. */
    public static class TransferException extends Exception {
        public TransferException(String message) {
            super(message);
        }
    }

    /** Deflate brut, pour qu'un carnet de cinquante entrees tienne dans un QR. */
    static byte[] deflate(byte[] raw) {
        Deflater deflater = new Deflater(Deflater.BEST_COMPRESSION);
        deflater.setInput(raw);
        deflater.finish();

        ByteArrayOutputStream out = new ByteArrayOutputStream();
        byte[] chunk = new byte[8192];
        while (!deflater.finished()) {
            out.write(chunk, 0, deflater.deflate(chunk));
        }
        deflater.end();
        return out.toByteArray();
    }

    static byte[] inflate(byte[] raw) throws TransferException {
        Inflater inflater = new Inflater();
        inflater.setInput(raw);

        ByteArrayOutputStream out = new ByteArrayOutputStream();
        byte[] chunk = new byte[8192];
        try {
            while (!inflater.finished()) {
                int n = inflater.inflate(chunk);
                if (n == 0 && (inflater.needsInput() || inflater.needsDictionary())) break;
                out.write(chunk, 0, n);
            }
        } catch (DataFormatException e) {
            throw new TransferException("Contenu illisible apres dechiffrement.");
        } finally {
            inflater.end();
        }
        return out.toByteArray();
    }

    /** Chiffre un carnet en un payload transportable. */
    public static String exportVault(@NonNull Vault vault, @NonNull String masterKey)
            throws GeneralSecurityException, JSONException {
        return exportJson(vault.toCompactJson(), masterKey);
    }

    /** Chiffre un carnet déjà sérialisé. Séparé pour les tests d'interopérabilité. */
    static String exportJson(@NonNull String json, @NonNull String masterKey)
            throws GeneralSecurityException {
        byte[] compressed = deflate(json.getBytes(StandardCharsets.UTF_8));

        byte[] nonce = new byte[NONCE_BYTES];
        RANDOM.nextBytes(nonce);
        Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
        cipher.init(Cipher.ENCRYPT_MODE, deriveKey(masterKey),
                new GCMParameterSpec(TAG_BITS, nonce));

        return PREFIX + "." + Base64Url.encode(nonce) + "."
                + Base64Url.encode(cipher.doFinal(compressed));
    }

    /** Dechiffre un payload. Leve TransferException s'il est illisible. */
    public static Vault importVault(@NonNull String payload, @NonNull String masterKey)
            throws TransferException {
        String[] parts = payload.trim().split("\\.");
        if (parts.length != 3) {
            throw new TransferException("Format inattendu : TC1.<nonce>.<donnees> attendu.");
        }
        if (!PREFIX.equals(parts[0])) {
            // Interpreter un format inconnu au hasard serait pire que refuser.
            throw new TransferException(
                    "Version « " + parts[0] + " » inconnue, ce client lit " + PREFIX + ".");
        }

        byte[] plain;
        try {
            plain = openBytes(deriveKey(masterKey),
                    Base64Url.decode(parts[1]), Base64Url.decode(parts[2]));
        } catch (GeneralSecurityException | IllegalArgumentException e) {
            throw new TransferException(
                    "Dechiffrement impossible : clef maitresse differente, ou donnees alterees.");
        }

        try {
            return Vault.fromJson(new String(inflate(plain), StandardCharsets.UTF_8));
        } catch (JSONException e) {
            throw new TransferException("Carnet illisible : " + e.getMessage());
        }
    }
}

File: shared/apple/Transfer.swift
//
//  Transfer.swift
//  Chiffrement du carnet avant qu'il ne quitte l'appareil.
//
//  Le serveur ne reçoit que des blocs opaques : il ne peut lire ni les sites,
//  ni les identifiants.
//
//  Le sel diffère de celui des mots de passe et de celui de l'empreinte : une
//  même valeur dérivée ne doit jamais servir à deux usages, sinon une faiblesse
//  sur l'un exposerait l'autre.
//
//  Source unique : shared/apple/Transfer.swift
//  Spécification : shared/spec/vault-transfer.md
//

import CommonCrypto
import Compression
import CryptoKit
import Foundation

public enum Transfer {

    public static let prefix = "TC1"

    private static let salt = "thecode-transfer/v1"
    private static let iterations: UInt32 = 600_000
    private static let keyBytes = 32

    public enum TransferError: Error, Equatable {
        /// Version inconnue, ou payload mal forme.
        case unreadable(String)
        /// La clef maîtresse n'est pas celle qui a servi à chiffrer, ou les
        /// données ont été altérées. AES-GCM ne distingue pas les deux, et
        /// c'est volontaire.
        case cannotOpen
        case derivationFailed
    }

    /// Dérive la clef de transfert depuis la clef maîtresse.
    public static func deriveKey(_ masterKey: String) throws -> SymmetricKey {
        var output = [UInt8](repeating: 0, count: keyBytes)
        let saltBytes = Array(salt.utf8)

        let status = masterKey.withCString { keyPointer in
            CCKeyDerivationPBKDF(
                CCPBKDFAlgorithm(kCCPBKDF2),
                keyPointer,
                strlen(keyPointer),
                saltBytes,
                saltBytes.count,
                CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA256),
                iterations,
                &output,
                keyBytes
            )
        }

        guard status == kCCSuccess else { throw TransferError.derivationFailed }
        return SymmetricKey(data: Data(output))
    }

    /// Un bloc chiffré : nonce et données, tous deux opaques pour le serveur.
    public struct Sealed {
        public let nonce: Data
        public let blob: Data
    }

    public static func seal(_ plain: Data, with key: SymmetricKey) throws -> Sealed {
        // AES.GCM.Nonce() tire un nonce aléatoire : le réutiliser avec la même
        // clef casserait AES-GCM.
        let nonce = AES.GCM.Nonce()
        let box = try AES.GCM.seal(plain, using: key, nonce: nonce)
        return Sealed(nonce: Data(nonce), blob: box.ciphertext + box.tag)
    }

    public static func open(nonce: Data, blob: Data, with key: SymmetricKey) throws -> Data {
        guard blob.count > 16 else { throw TransferError.cannotOpen }

        do {
            let box = try AES.GCM.SealedBox(
                nonce: AES.GCM.Nonce(data: nonce),
                ciphertext: blob.prefix(blob.count - 16),
                tag: blob.suffix(16)
            )
            return try AES.GCM.open(box, using: key)
        } catch {
            throw TransferError.cannotOpen
        }
    }

    // MARK: - Découpage en plusieurs QR

    /// Un QR plafonne à ~2,9 Ko. On garde de la marge pour l'en-tête du
    /// fragment, qui s'ajoute à chaque morceau.
    public static let fragmentPayloadLimit = 2600

    /// Découpe un payload en fragments affichables l'un après l'autre.
    ///
    /// Un seul fragment quand le payload tient : inutile d'imposer un
    /// assemblage pour un carnet ordinaire.
    public static func fragments(_ payload: String) -> [String] {
        guard payload.count > fragmentPayloadLimit else { return [payload] }

        // Le préfixe « TC1. » est porté une fois par le réassemblage, pas par
        // chaque fragment.
        let body = String(payload.dropFirst(prefix.count + 1))
        let chunks = stride(from: 0, to: body.count, by: fragmentPayloadLimit).map { start -> String in
            let from = body.index(body.startIndex, offsetBy: start)
            let to = body.index(from, offsetBy: min(fragmentPayloadLimit, body.count - start))
            return String(body[from..<to])
        }

        return chunks.enumerated().map { index, chunk in
            "TC1m.\(index).\(chunks.count).\(chunk)"
        }
    }

    // MARK: - Carnet entier

    /// Chiffre un carnet en un payload transportable.
    public static func exportVault(_ vault: Vault, masterKey: String) throws -> String {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        let compressed = try deflate(encoder.encode(vault))

        let sealed = try seal(compressed, with: deriveKey(masterKey))
        return "\(prefix).\(Base64URL.encode(sealed.nonce)).\(Base64URL.encode(sealed.blob))"
    }

    /// Déchiffre un payload. Lève `TransferError` s'il est illisible.
    public static func importVault(_ payload: String, masterKey: String) throws -> Vault {
        let parts = payload.trimmingCharacters(in: .whitespacesAndNewlines).split(
            separator: ".", omittingEmptySubsequences: false
        ).map(String.init)

        guard parts.count == 3 else {
            throw TransferError.unreadable("Format inattendu : TC1.<nonce>.<donnees> attendu.")
        }
        guard parts[0] == prefix else {
            // Interpréter un format inconnu au hasard serait pire que refuser.
            throw TransferError.unreadable(
                "Version « \(parts[0]) » inconnue, ce client lit \(prefix).")
        }
        guard let nonce = Base64URL.decode(parts[1]), let blob = Base64URL.decode(parts[2]) else {
            throw TransferError.unreadable("Encodage invalide.")
        }

        let compressed = try open(nonce: nonce, blob: blob, with: deriveKey(masterKey))
        return try JSONDecoder().decode(Vault.self, from: try inflate(compressed))
    }
}

// MARK: - Compression zlib

/// Compresse, pour qu'un carnet de cinquante entrées tienne dans un QR.
///
/// Le format attendu est **zlib** (RFC 1950) : c'est ce que produisent
/// `zlib.compress` côté Python, `Deflater` côté Java et `CompressionStream
/// ("deflate")` côté navigateur — où « deflate » désigne justement le format
/// zlib, « deflate-raw » étant le format brut.
///
/// `COMPRESSION_ZLIB` d'Apple, malgré son nom, produit du deflate **brut**
/// (RFC 1951). On ajoute donc l'en-tête et l'Adler-32 nous-mêmes. Sans cela
/// l'aller-retour local fonctionne — et rien d'autre ne relit le payload.
private func deflate(_ data: Data) throws -> Data {
    var out = Data([0x78, 0x9C])
    out.append(try perform(data, operation: COMPRESSION_STREAM_ENCODE))

    var checksum = adler32(data).bigEndian
    out.append(Data(bytes: &checksum, count: 4))
    return out
}

private func inflate(_ data: Data) throws -> Data {
    // 2 octets d'en-tête, 4 d'Adler-32 : le reste est du deflate brut.
    guard data.count > 6, data[data.startIndex] & 0x0F == 8 else {
        throw Transfer.TransferError.unreadable("Contenu compressé illisible.")
    }
    let body = data.dropFirst(2).dropLast(4)
    let plain = try perform(Data(body), operation: COMPRESSION_STREAM_DECODE)

    let expected = data.suffix(4).reduce(UInt32(0)) { ($0 << 8) | UInt32($1) }
    guard adler32(plain) == expected else {
        throw Transfer.TransferError.unreadable("Somme de contrôle invalide.")
    }
    return plain
}

/// Adler-32, tel que le définit RFC 1950. Trivial, et il évite d'accepter un
/// contenu décompressé de travers.
private func adler32(_ data: Data) -> UInt32 {
    var a: UInt32 = 1
    var b: UInt32 = 0
    for byte in data {
        a = (a + UInt32(byte)) % 65521
        b = (b + a) % 65521
    }
    return (b << 16) | a
}

private func perform(_ data: Data, operation: compression_stream_operation) throws -> Data {
    guard !data.isEmpty else { return Data() }

    var stream = compression_stream(
        dst_ptr: UnsafeMutablePointer<UInt8>(bitPattern: 1)!, dst_size: 0,
        src_ptr: UnsafePointer<UInt8>(bitPattern: 1)!, src_size: 0, state: nil)
    guard compression_stream_init(&stream, operation, COMPRESSION_ZLIB) == COMPRESSION_STATUS_OK
    else {
        throw Transfer.TransferError.unreadable("Compression indisponible.")
    }
    defer { compression_stream_destroy(&stream) }

    let bufferSize = 32_768
    let buffer = UnsafeMutablePointer<UInt8>.allocate(capacity: bufferSize)
    defer { buffer.deallocate() }

    var output = Data()
    let flags = Int32(COMPRESSION_STREAM_FINALIZE.rawValue)

    try data.withUnsafeBytes { (raw: UnsafeRawBufferPointer) in
        stream.src_ptr = raw.bindMemory(to: UInt8.self).baseAddress!
        stream.src_size = data.count

        repeat {
            stream.dst_ptr = buffer
            stream.dst_size = bufferSize

            switch compression_stream_process(&stream, flags) {
            case COMPRESSION_STATUS_OK, COMPRESSION_STATUS_END:
                output.append(buffer, count: bufferSize - stream.dst_size)
                if stream.dst_size != 0 { return }
            default:
                throw Transfer.TransferError.unreadable("Contenu illisible après déchiffrement.")
            }
        } while true
    }

    return output
}
